i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,248 @@
+        self.with_pos_bias = cfg.with_pos_bias
+        self.pos_emb_dropout = nn.Dropout(cfg.pos_emb_dropout)
+
+        assert not self.learnable_pos_emb or self.rel_pos_clip
+
+        self.att_weights_dropout = nn.Dropout(cfg.att_weights_dropout)
+
+        assert self.embed_dim % self.num_heads == 0, "embed_dim must be divisible by num_heads"
+
+        # projection matrices
+        self.qkv_proj = nn.Linear(self.embed_dim, 3 * self.embed_dim, bias=cfg.with_bias)
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=cfg.with_bias)
+
+        self.register_parameter("rel_pos_embeddings", None)
+        self.register_parameter("pos_bias_u", None)
+        self.register_parameter("pos_bias_v", None)
+
+        self.pos_emb_dim = (
+            self.embed_dim if cfg.with_linear_pos or cfg.separate_pos_emb_per_head else self.embed_dim_per_head
+        )
+        if self.learnable_pos_emb:
+            self.rel_pos_embeddings = nn.parameter.Parameter(torch.empty(self.rel_pos_clip * 2 + 1, self.pos_emb_dim))
+        if cfg.with_linear_pos:
+            self.linear_pos = nn.Linear(
+                self.pos_emb_dim,
+                self.embed_dim if cfg.separate_pos_emb_per_head else self.embed_dim_per_head,
+                bias=False,
+            )
+        else:
+            self.linear_pos = nn.Identity()
+
+        if self.with_pos_bias:
+            self.pos_bias_u = nn.parameter.Parameter(torch.empty(self.num_heads, self.embed_dim_per_head))
+            self.pos_bias_v = nn.parameter.Parameter(torch.empty(self.num_heads, self.embed_dim_per_head))
+
+        self.dropout = BroadcastDropout(cfg.dropout, dropout_broadcast_axes=cfg.dropout_broadcast_axes)
+
+        self._reset_parameters()
+
+    def _reset_parameters(self):
+        if self.learnable_pos_emb:
+            nn.init.xavier_normal_(self.rel_pos_embeddings)
+        if self.with_pos_bias:
+            # init taken from espnet default
+            nn.init.xavier_uniform_(self.pos_bias_u)
+            nn.init.xavier_uniform_(self.pos_bias_v)
+
+    def forward(self, input_tensor: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
+        """
+        Apply layer norm and multi-head self attention and dropout
+
+        :param input_tensor: Input to the self attention of shape (B, T, F)
+        :param sequence_mask: bool mask of shape (B, T), True signals within sequence, False outside
+        """
+        output_tensor = self.layernorm(input_tensor)  # [B, T, F]
+
+        time_dim_size = output_tensor.shape[1]
+        batch_dim_size = output_tensor.shape[0]
+
+        # attention mask
+        # T: query seq. length, T' key/value seg length; T = T' if same input tensor
+        inv_sequence_mask = compat.logical_not(sequence_mask)  # [B, T']
+        mask = (
+            torch.zeros_like(inv_sequence_mask, dtype=input_tensor.dtype)
+            .masked_fill(inv_sequence_mask, float("-inf"))
+            .view(batch_dim_size, 1, 1, time_dim_size)
```

Comment:
```suggestion
            .reshape(batch_dim_size, 1, 1, time_dim_size)
```
(Btw, in general, you should almost never use `view`, unless you have a very good reason for it.)

Reply:
why? I thought view() could ensure no extra copy is done and save some memory.

Reply:
okay, reshape could do the same thing.

Reply:
`view` fails if it cannot do it, while `reshape` always works, and does the same saving of memory if possible.